### PR TITLE
Feature: RSpec Part 2: Hooks, Let and Subject

### PR DESCRIPTION
**Pull request by KevinMulhern**
Because:
* We are rewriting the testing lessons in the Ruby course and need a lesson to teach the student about coding sharing techniques in RSpec.

This lesson covers:
* Before and after hooks
* Let variables
* Subject
* Context groups

### CODE REVIEW

File: ruby_programming/automated_testing/rspec_part_two_code_sharing.md [hunk 1]
```diff
@@ -0,0 +1,525 @@
+### Introduction
+
+In this lesson we will be exploring some of the tools RSpec provides for sharing code among multiple tests. Specifically, we will be looking at hook, let and subject helper methods. These methods are frequently used in RSpec test suites so it's important to how they work and when to use them.
```
Comment (CouchofTomato) [inline ruby_programming/automated_testing/rspec_part_two_code_sharing.md:3]: >so it's important to how they work and when to use them.

Missing the word 'know'

File: ruby_programming/automated_testing/rspec_part_two_code_sharing.md [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+### Let and Subject Variables
+
+Using instance variables in before hooks like we have been doing has some big drawbacks that can lead to slower and less stable test suites.
+
+- Before hooks are run for every test in the same example group. Some of those tests may not need that instance variable.
+- Instance variables return nil if they are not assigned, it's fairly easy and a common mistake to misspell an instance variable and end up with a no method error for nil class in Ruby. You've probably seen this error a few times already while working through the OOP section.
+- Instance variables can leak between tests and even between test files. Which could potentially cause all sorts of havoc in our test suite.
+
+Thankfully RSpec provides a more idiomatic and and safer way of creating variables that can be used in multiple tests, The `let` method.
```
Comment (CouchofTomato) [inline ruby_programming/automated_testing/rspec_part_two_code_sharing.md:204]: double 'and' here

**Review by CouchofTomato — APPROVED** [review]
Great work again mate. Awesome stuff.

**Review by rlmoser99 — CHANGES_REQUESTED** [review]
Another amazing job on this lesson! 🎉 

Just found one thing as I was following along that you probably refactored at some point.

File: ruby_programming/automated_testing/rspec_part_two_code_sharing.md [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+# lib/user.rb
+
+class User
+  attr_reader :name, :email, :age
+
+  def initialize(name:, email:, age:)
+    @name = name
+    @email = email
+    @age = age
+  end
```
[the remark this answers is not in the document]
Comment (KevinMulhern) [reply ruby_programming/automated_testing/rspec_part_two_code_sharing.md]: Thanks @rlmoser99, I was using keyword args initially and must have forgot to update this example 🙈

**Pull request merged**